### ISSUE TITLE
clustering: noise-reduction roadmap — quality filter -> UMAP -> tuning to break <20% noise

### ISSUE THREAD

**Issue by theGaryLarson** [Building-With-Agents/job-intelligence-engine#327]
## Goal — many fine-grained canonical roles, not few mega-clusters

**Target: 20+ canonical roles, mega-cluster (largest single cluster) <10 % of corpus.**

The earlier Tier 1/2 work was implicitly biased toward HDBSCAN's `eom` default — "find the few most stable clusters" — which is what gave us the 1,000-post "Senior SRE" mega-cluster. The actual product goal is the opposite: a richer canonical-role catalog that resolves "software developer", "data engineer", "DevOps engineer", "ML engineer" etc. as **separate** roles, not a merged super-cluster.

This re-framing changes the optimal HDBSCAN parameter choices, documented inline below. Phase 2 tuning priorities reorder accordingly.

---

## Context

Tier 1 (#317) and Tier 2 (#321/#325) cosine-distance HDBSCAN runs both fail the smoke target of <20% noise AND fail the new "many roles" goal:

| Run | min_cluster_size | min_samples | sel_method | Clusters | Noise | Mega-cluster share |
|---|---:|---:|---|---:|---:|---:|
| Pre-cosine (euclidean) | 10 | 5 | eom | 3 | 28.5% | ~94% |
| Tier 1 (cosine) | 10 | 5 | eom | 9 | **67.6%** | ~26% |
| Tier 2 (cosine + tightening) | 5 | 10 | eom | 11 | **73.2%** | ~26% |
| **Target** | — | — | **leaf** | **≥20** | **<30%** | **<10%** |

Tier 2 made noise *worse*, confirming the curse-of-dimensionality residual: 1536-D Azure embeddings have too-uniform pairwise distances for density-based clustering even with cosine + L2 normalization.

This issue tracks the **roadmap to reach the "many canonical roles" target** — multi-phase, one variable at a time.

## Quality-score baseline (collected 2026-04-29 against live DB)

```
total                    4,711
NULL quality_score         460  (9.8%)  -> issue #328
range                  0.183 - 0.75              -> issue #329 (upper half empty)
mean                     0.498
quality_score >= 0.5   2,382  (50.6%)
quality_score >= 0.6     160  (3.4%)
quality_score >= 0.7       8  (0.17%)
```

The compressed score range is being investigated separately in #329; the NULL set in #328. **Until those land, do not apply a `quality_score` filter to clustering input** — the scorer's signal is too compressed to be a reliable lever, and any threshold we pick now will be re-tuned once #329 ships.

---

## Roadmap — execute phases sequentially

> **Discipline rule (per #297):** ONE variable change per re-run. Each Azure-embedding pass costs ~4 minutes + real $; do not waste runs on multi-lever experiments. After each phase, write a findings note in `eval/runs/findingsv2.X-clustering-tierN.md` covering: noise rate, cluster count, top-cluster member share, qualitative cluster-label review (sample 5 postings per cluster).

### Phase 1 — UMAP dimensionality reduction (the textbook fix)

**Hypothesis:** Reducing 1536-D → 15-D via UMAP restores meaningful density gradients so HDBSCAN can find clusters. This is the BERTopic-standard pipeline; expected noise drop to 10–25%. **This is the highest-leverage lever; doing it first lets all downstream tuning happen in low-D where parameters mean what the docs say.**

**Implementation:**
- `requirements.txt` — add `umap-learn==0.5.6` (or current pinned).
- `config/clustering.yaml` — add a new `umap:` block AND adjust HDBSCAN params toward the "many clusters" goal:

```yaml
clustering:
  # UMAP block — gated by feature flag for reversibility
  umap:
    enabled: true              # set false to skip and run HDBSCAN on raw 1536-D
    n_components: 15           # target dim — more dims = more separation between similar roles
    n_neighbors: 10            # local neighborhood size — smaller = preserve local structure (more clusters)
    min_dist: 0.0              # tight clusters required for density methods
    metric: cosine
    random_state: 42           # reproducibility

  # HDBSCAN params — biased for many clusters, not few stable ones
  min_cluster_size: 5
  min_samples: 5                  # REVERT from Tier 2's 10; less conservative = more clusters
  cluster_selection_epsilon: 0.0  # do NOT raise — epsilon MERGES clusters, opposite of our goal
  cluster_selection_method: leaf  # KEY CHANGE — was eom; leaf returns more, smaller clusters with less noise
  alpha: 1.0                      # do NOT raise to 1.3 — that's "conservative" (fewer clusters)
  distance_metric: cosine
```

- `analytics/clustering/pipeline.py` — between `embed_posting_features` and `run_clustering`, conditionally apply UMAP:

```python
if cluster_umap_enabled():
    reducer = umap.UMAP(
        n_components=cluster_umap_n_components(),
        n_neighbors=cluster_umap_n_neighbors(),
        min_dist=cluster_umap_min_dist(),
        metric=cluster_umap_metric(),
        random_state=cluster_umap_random_state(),
    )
    matrix = reducer.fit_transform(matrix)
    # After UMAP, switch HDBSCAN metric to euclidean; cosine in low-D is
    # equivalent and algorithm="best" becomes available for speed.
```

Also wire `cluster_selection_method` through to the HDBSCAN constructor (it currently isn't a config-driven param — adding it is part of this PR).

- `analytics/clustering/config.py` — add `DEFAULT_CLUSTER_UMAP_*` constants + accessors AND `DEFAULT_CLUSTER_SELECTION_METHOD = "leaf"` + accessor. All entries land in the parity test automatically.
- Real-instantiation guard test in `analytics/tests/test_clustering_package.py` (mirror the pattern from #319's `test_hdbscan_accepts_cosine_generic_kwargs`): instantiate real `umap.UMAP(...).fit_transform(...)` on a tiny matrix to catch API changes / version skew.
- Live re-run + fixture PR.

**Measurements to report:**
- `eligible_posting_count` (still 4,733 — no quality filter yet)
- noise_rate
- cluster count
- mega-cluster member share (currently 26%)
- per-cluster top skills/tools — verify UMAP did not destroy semantic coherence
- UMAP convergence diagnostics (embedding shape, any warnings)

**Decision gate:**
- noise <30% AND cluster_count ≥ 20 AND mega-cluster share <10% → ship; advance to Phase 2 for fine-tuning
- noise <30% but cluster_count <20 → Phase 2, prioritize levers that increase cluster count
- noise >30% → revisit UMAP params (n_components 10/15/20 sweep, n_neighbors 8/10/15 sweep) before Phase 2

**Effort:** ~1.5–2 days. Pair C — joint design session for the pipeline edit; Bryan owns the YAML + accessors, Emilio owns the pipeline integration + guard test.

---

### Phase 2 — Tuning sweep on the UMAP-reduced space

**Reordered for the "many canonical roles" goal.** The previous order was tuned to "reduce noise"; this order is tuned to "raise cluster count without sacrificing coherence."

**Tuning matrix (one variable change per run, in this order):**

| Priority | Knob | Default → Try | Why this order |
|---:|---|---|---|
| 1 | `min_samples` | 5 → 3 → 7 | Lowest = most permissive = more clustered points + more clusters survive. Direct from `parameter_selection.rst` docs: `min_samples=1` is the "liberal" extreme. |
| 2 | `min_cluster_size` | 5 → 3 → 8 | Smaller floor = more candidate clusters survive selection. Already at 5; testing 3 is the next aggressive step. |
| 3 | UMAP `n_components` | 15 → 10 → 20 → 30 | More dims preserves more semantic distinction → more distinguishable roles. 30 risks curse of dim returning. |
| 4 | UMAP `n_neighbors` | 10 → 5 → 15 | Lower = more local structure → more clusters. Fine-tune only after the above settle. |
| 5 | **Soft clustering / noise reassignment** (`hdbscan.all_points_membership_vectors`) | not used → enabled | Apply post-hoc to noise points. **Does not change cluster count; raises effective coverage by giving every noise point a probability over the canonical roles**. argmax gives a meaningful assignment. |
| 6 | `cluster_selection_epsilon` | 0.0 → consider lowering effective threshold | **Only consider if we have TOO MANY clusters (>40) and want to merge.** Raising epsilon REDUCES cluster count — opposite direction from levers 1-4. |

**Discipline:** Run them in priority order. Stop when target is met (cluster_count ≥ 20, mega-cluster <10%, noise <30%); do not stack changes. Log each iteration in a section of the Phase 2 findings doc.

**Effort:** ~1–2 days, depending on how many tuning iterations are needed.

---

### Phase 3 — Sub-cluster the residual mega-cluster (BranchDetector or two-pass)

**Hypothesis:** Even with optimized Phase 1+2 settings, one or two clusters may remain disproportionately large because they contain semantically distinct sub-roles (e.g., "Senior SRE" might really be DevOps + Platform Engineer + Cloud Architect). Decompose them.

**Two options to evaluate:**

**Option A — `hdbscan.BranchDetector`** ([upstream docs](https://github.com/scikit-learn-contrib/hdbscan/blob/master/docs/how_to_detect_branches.rst)):

```python
clusterer = hdbscan.HDBSCAN(
    min_cluster_size=5,
    cluster_selection_method="leaf",
    branch_detection_data=True,
    prediction_data=True,
).fit(reduced_matrix)

branch_detector = hdbscan.BranchDetector(
    min_cluster_size=3,
    branch_detection_method="full",
).fit(clusterer)
# branch_detector.labels_ now has finer-grained subdivisions
```

**Option B — Two-pass clustering:**
- Pass 1: HDBSCAN with `eom` to find macro structure (gives the few mega-clusters back)
- Pass 2: For any cluster whose member share > 10 %, run HDBSCAN again on JUST those postings with `leaf` and aggressive params to subdivide

**Which to ship:** Try BranchDetector first (single-pass, simpler). Fall back to two-pass only if BranchDetector doesn't give clean splits.

**Effort:** ~1 day if Phase 1+2 already hit target; up to 2 days if multiple mega-clusters need splitting.

---

### Phase 4 — Quality-score pre-filter (DEFERRED — depends on #328 + #329)

> **Do not start Phase 4 until both [#328](https://github.com/Building-With-Agents/job-intelligence-engine/issues/328) and [#329](https://github.com/Building-With-Agents/job-intelligence-engine/issues/329) close.**

Once the Enrichment Agent's quality-score logic is reviewed (#329) and the NULL-score gap is closed (#328), revisit whether a quality filter on the clustering input is a useful lever. The threshold must be picked AFTER the scorer's distribution is repaired — picking 0.5 today against the compressed [0.18, 0.75] range would be arbitrary.

**Implementation (when unblocked):**
- `config/clustering.yaml` — add `cluster_min_quality_score: <TBD post-#329>`.
- `analytics/clustering/config.py` — add `DEFAULT_CLUSTER_MIN_QUALITY_SCORE` + accessor (mirror the parity pattern from #325).
- `analytics/canonical_roles/loader.py:50` — extend the SQL `WHERE` clause: `AND (jp.quality_score IS NOT NULL AND jp.quality_score >= :min_quality_score)`. Threshold reads from the new accessor.
- `analytics/tests/test_clustering_config_parity.py` — add the new constant.
- Live re-run + fixture PR.

**Decision gate:** Phase 4 is value-add, not gating. If Phase 1+2+3 already meet the target, Phase 4 is optional.

**Effort:** ~0.5 day once unblocked.

---

### Phase 5 — Lock in winning config + close out

- Update `config/clustering.yaml` + `analytics/clustering/config.py` with the final winning values.
- Update parity test (auto-passes if new accessors / new defaults follow the pattern).
- Final findings doc: `eval/runs/findingsv2.X-clustering-tier3-final.md`. Cover: full parameter trace, what each knob did, what the cluster catalog looks like.
- Re-run smoke test — must pass `noise <30%`, `canonical_roles >= 20`, mega-cluster <10%, and `"software developer" Q&A` resolves.
- Update [scripts/smoke/e2e_clustering_cosine.py](scripts/smoke/e2e_clustering_cosine.py) thresholds:
  - `noise_fraction < 30%` (was 20% — relaxed because many small clusters legitimately have more boundary noise)
  - `canonical_roles count >= 20` (was 10 — raised to match the new goal)
- Close #321; close this issue if smoke passes.

If smoke STILL fails after Phase 5 → file a data-level investigation issue (not an algorithm one). At that point we have exhausted reasonable algorithm levers; remaining noise is real signal.

---

## Discipline + ownership

- **One variable per re-run.** Cite the changed knob in the commit message.
- **Each phase ends with a findings doc + a one-file fixture PR.** Same pattern as PR #320 / #326.
- **@theGaryLarson owns live DB re-runs.** Pair C builds + tests locally; Gary triggers the re-cluster against source-of-truth and pushes the fixture PR.
- **No alternative algorithms** (k-means, agglomerative) without explicit team-lead sign-off. UMAP + HDBSCAN (with optional BranchDetector or two-pass for sub-clustering) is the planned terminal architecture.
- **Don't add `cluster_selection_epsilon > 0`** in Phase 2 unless we explicitly have too many clusters and need to merge. Epsilon is the wrong direction for the "many roles" goal.

## Acceptance criteria (issue closes when all checked)

- [ ] Phase 1 UMAP shipped (with `cluster_selection_method=leaf`) + findings doc + fixture PR merged
- [ ] Phase 2 tuning sweep complete + best config locked in (≥ 20 clusters target)
- [ ] Phase 3 mega-cluster decomposition (BranchDetector or two-pass) shipped — OR explicitly skipped with rationale
- [ ] #328 + #329 closed (prerequisite for Phase 4)
- [ ] Phase 4 quality filter shipped (or explicitly skipped with rationale) + findings doc
- [ ] Phase 5 final findings doc committed
- [ ] Smoke test thresholds updated to `noise < 30%`, `canonical_roles >= 20`, mega-cluster <10%
- [ ] `python scripts/smoke/e2e_clustering_cosine.py` passes against the new thresholds

## References

- #297 — parent clustering quality issue with the original Tier plan
- #321 — Tier 2 (now superseded by this roadmap)
- #328 — investigate NULL quality_score on jsearch postings (Phase 4 prerequisite)
- #329 — investigate quality_score [0.18, 0.75] compression (Phase 4 prerequisite)
- PR #325 — Tier 2 config (merged)
- PR #326 — Tier 2 fixture regen
- HDBSCAN docs: [`parameter_selection.rst`](https://github.com/scikit-learn-contrib/hdbscan/blob/master/docs/parameter_selection.rst), [`how_to_detect_branches.rst`](https://github.com/scikit-learn-contrib/hdbscan/blob/master/docs/how_to_detect_branches.rst), [`soft_clustering.rst`](https://github.com/scikit-learn-contrib/hdbscan/blob/master/docs/soft_clustering.rst)
- `eval/runs/findingsv2.X-*` — running findings docs

**Comment by BryanPMX** [comment]
## Status update — (Pair C / Bryan)

Here is the current state of the roadmap against the acceptance criteria:

### Phase 1 — UMAP (shipped ✅)

UMAP + `cluster_selection_method=leaf` + `min_samples` revert (10→5) landed in `development`. `findingsv2.5` captures the live-run measurements:

| Metric | Before (Tier 2) | Phase 1 | #327 target |
|---|---:|---:|---:|
| Noise rate | 73.2% | **36.5%** | <30% (off by 6.5pp) |
| Clusters | 11 | **200** | ≥20 ✅ |
| Mega-cluster share | ~26% | **1.1%** | <10% ✅ |

Cluster count and mega-cluster targets cleared; noise landed 6.5pp over target with Phase 1 alone.

### Phase 2 — Tuning sweep (in progress)

- **`min_samples` 5→3** — config shipped to `development` (YAML + `analytics/clustering/config.py`); live re-run pending Gary. Placeholder in `findingsv2.8` waiting for measurements.
- **`min_cluster_size` 5→3** (PR #402, Nestor) — approved today; CI green; branch needs `git merge development` before merge. Nestor's findings doc (`eval/runs/findings-400-hdbscan-min-cluster-size-3.md`) shows noise **47.7% → 31.4%** (just above the 30% gate). Once both knobs are live, Gary runs the combined re-cluster — expect the combined effect to clear <30%.

### Phase 3 — BranchDetector (skipped ✅)

Explicitly skipped with rationale in `findingsv2.9`: no mega-cluster ≥10% after Phase 1+2, so there's nothing to decompose. Re-open criteria documented (mega ≥10% after Phase 2 lock, or product request).

### Phase 4 — Quality filter (unblocked, not yet activated)

Both prerequisites are now closed: #328 (NULL quality_score fix) and #329 (score compression investigation). The gate code is in place at `analytics/canonical_roles/loader.py` and `config/clustering.yaml` (`cluster_input_min_quality_score: 0.0` = disabled). Needs a threshold decision after the Phase 2 baseline solidifies — no urgency to activate until Phase 2 noise measurements are in.

### Phase 5 — Final config + smoke thresholds (shipped ✅)

`findingsv2.11` records the accepted thresholds:
- `noise_fraction < 30%` (was 20%)
- `canonical_roles >= 20` (was 10)
- mega-cluster share `< 10%`

Smoke test code updated accordingly.

---

### Remaining open items before closing

1. **PR #402 merge** — Nestor to update branch (`git merge development`), re-push; Gary to run the live re-cluster and paste measurements into `findingsv2.8`.
2. **Confirm post-402 smoke passes** — `python scripts/smoke/e2e_clustering_cosine.py` against the updated config (noise target now <30%, so 31.4% from #402's finding should clear it after the combined `min_samples=3` + `min_cluster_size=3` run).
3. **Close this issue** — once smoke passes on the combined config.
